### ISSUE TITLE
PiAware will not build on Ubuntu 24.04.

### ISSUE THREAD

**Issue by jprochazka** [jprochazka/adsb-receiver#575]
### NOTE: This issue affects any release using Python 3.12 or greater not just Nobel.

The PiAware build halts with the following error when trying to build the PiAware package.

```
  File "/home/joe/adsb-receiver/build/piaware_builder/package-bookworm/debian/venv/lib/python3.12/site-packages/cx_Freeze/finder.py", line 33, in <module>
    IMPORT_STAR = opcode.opmap["IMPORT_STAR"]
                  ~~~~~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 'IMPORT_STAR'
```

It would appear that cx_Freeze does not support Paython 3.12 which is the version of Python installed on Ubuntu Noble Numbat or the Ubuntu team has not yet updated things on their end. A fix for this issue appears to be available from the cx_Freeze developers but at this time the newer release supporting Python 3.12 is not available via the Ubuntu package repositories. I have been trying to implement a fix located in the following post from the cx_Freeze issue tracker using pipx and pip in a venv with no joy. 

https://github.com/marcelotduarte/cx_Freeze/issues/2153

Needless to say, the workaround is getting a bit more complicated than I would like it as far as this project is concerned. I am thinking that Noble Numbat support will need to be put on hold until a more permanent and stable solution is found.

**Update:**

After reading a post by @abcd567a on the FlightAware forums and looking through the bash scripts he kindly offered up to the community I found that it is possible to build PiAware on Nobel but it entails ditching MLAT functionality. IMO this is still a no go and until FlightAware team finds a fix or better yet the Ubuntu team gets on top updating cx_Freeze if it is in fact at a stable stage, Nobel support is still a no go unless anyone else has any ideas.

**Update Two:**

PiAware's MLAT code is using the asyncore module which was deprecated in version 3.6 and announced for removal in version 3.12 at the time of its release. Until the software is updated to use asyncio which is recommended in the Python documentation or some other solution it appears PiAware with MLAT support installations on systems using Python 3.12+ will not be possible.

https://peps.python.org/pep-0594/#asyncore

**Comment by abcd567a** [comment]
I have tried to resolve this issue in March 2024 with marcelotdurate (author of  cx_Freeze) and obj Oliver Jowet (author of mlat-client). Please see following thread in Flightaware forums:

https://discussions.flightaware.com/t/ubuntu-24-04-noble-numbat-desktop-server-install-piaware-dump1090-fa-dump978-fa-and-piaware-web/90939)

<Br>

![mlat-client-incompatability-with-python3 12](https://github.com/jprochazka/adsb-receiver/assets/28452511/d7b1a4f6-b292-4145-879a-13e044ab9e76)

<Br>

![mlat-client-asyncio-python3 12](https://github.com/jprochazka/adsb-receiver/assets/28452511/d6b06168-5e5e-4191-9b2b-28f80ee6adb1)

<Br>

**Comment by jprochazka** [comment]
Thanks for the information @abcd567a.
Looks the addition of Nobel support will be a matter of waiting on updates from third parties.

**Comment by abcd567a** [comment]
Yes, the work required is to update mlat-client source code to be compatible with python3.12 which uses `asyncio` instead of `asyncore` used by python3.11 and earlier.

Debian 13 Trixi Sed (still) uses python3.11, and hence Piaware OS Trixie will also use python3.11. As a result there seems to be no urgency for Flightaware to update mlat-client source code.

<Br>

![Screenshot_20240705_090830_Samsung Internet](https://github.com/jprochazka/adsb-receiver/assets/28452511/8c6c6b02-67d9-4e5a-be0c-73c2c88e8fcb)

**Comment by jprochazka** [comment]
Makes sense being the asyncore module was marked as deprecated in Python 3.6 back in 2016 when it was released and marked for removal in 3.12.

https://docs.python.org/3.11/library/asyncore.html

**Comment by jprochazka** [comment]
Makes sense being the asyncore module was marked as deprecated in Python 3.6 when it was released and marked for removal in 3.12. I am sure a fix will be incoming shortly, defiantly sometime before Trixie is released is my guess. :)

https://docs.python.org/3.11/library/asyncore.html

**Comment by jprochazka** [comment]
It looks as though someone packaged up the old asyncore module and made it available for those using Python 3.12+. I will begin work on support for Noble Numbat hopefully tonight. If things go well look for a release supporting Ubuntu 24.04 to be made available sometime this weekend.

[pyasyncore on PyPi](https://pypi.org/project/pyasyncore/)
[Github Repository](https://github.com/simonrob/pyasyncore)

“This package contains the [asyncore](https://docs.python.org/3.11/library/asyncore.html) module as found in Python versions prior to 3.12. It is provided so that existing code relying on import asyncore is able to continue being used without significant refactoring.”

**Comment by jprochazka** [comment]
Still encountering the `KeyError: 'IMPORT_STAR'` error as outlined originally.

**Comment by jprochazka** [comment]
I submitted a [pull request](https://github.com/flightaware/piaware_builder/pull/26) to the piaware_builder repository which should fix this issue. I have experienced success on my end utilizing these changes.

**Comment by jprochazka** [comment]
The master branch currently contains scripting to use my fork of piaware_builder I have experienced success with. This will be the case until either the mentioned pull request is accepted, or another fix is merged into the piaware_builder repository. PiAware Trixie and Noble PiAware installs using this fork should be considered as **use at your own risk** until a fix is merged into the official piaware_builder repository.

**_All other distributions use the official piaware_build repository when building and installing PiAware._**

**Comment by jprochazka** [comment]
@retornam added to my pull request for piaware_builder. These changes should be added to the fork being used. https://github.com/flightaware/piaware_builder/pull/30/changes/ca1342d6be4f2d0a5c71cd826b1c9a3e962725f

**Comment by jprochazka** [comment]
It looks as though the above mentioned pull request has been merged into the piaware_builder dev branch.

https://github.com/flightaware/piaware_builder/pull/31
https://github.com/flightaware/piaware_builder/pull/30
https://github.com/flightaware/piaware_builder/pull/26

For v2.9.10 I will work on utilizing said changes by switching to a branch the utilizes these changes before releasing it.

**Comment by jprochazka** [comment]
It looks like my fork has become outdated and may have issues with Debian Trixie. I am going to try modifying the update branch to use the official piaware_builder dev branch and see what happens.

**Comment by jprochazka** [comment]
_I keep hearing the piaware_builder dev branch along with cx_Freeze 8.x will not work on Ubuntu 24.04 Noble. I have been told cx_Freeze 8.x will not work with Python 3.12 either. But I am also told that PiAware does not work on any version of Ubuntu as well but we all know this is not true_

Anyways I got the official piaware_builder dev branch to build no problem unmodified right out of the box and then again with a minor unnecessary one to silence warnings which involves setuptools exclusions. All I need to do is test it against live data to make sure MLAT is working.

The piaware.sh bash script needs to be pointed to the dev branch of the official FlightAware repository as well.

I added the following to the bottom of cx_Freeze-8.5.0.patch. (setuptools noise fix)

```
@@ -111,7 +111,7 @@
 version = {attr = "cx_Freeze.__version__"}
 
 [tool.setuptools.packages.find]
-exclude = ["samples", "tests*"]
+exclude = ["samples", "tests*", "cx_Freeze.importshed*"]
 namespaces = false
```

**Comment by abcd567a** [comment]
Piaware did built on Ubuntu 24, and mlat works. Built using dev branch of piaware_builder which uses cx_Freeze ver 8.5.0

```
pi@ubuntu-noble:~$ sudo journalctl -u piaware -n 8  

Apr 07 11:09:15 ubuntu-noble piaware[1081]: 2101 msgs recv'd from dump1090-fa (1891 in last 5m); 2101 msgs sent to FlightAware
Apr 07 11:14:13 ubuntu-noble piaware[1081]: 4007 msgs recv'd from dump1090-fa (1906 in last 5m); 4007 msgs sent to FlightAware
Apr 07 11:18:45 ubuntu-noble piaware[1081]: mlat-client(2096): Receiver status: connected
Apr 07 11:18:45 ubuntu-noble piaware[1081]: mlat-client(2096): Server status:   synchronized with 291 nearby receivers
Apr 07 11:18:45 ubuntu-noble piaware[1081]: mlat-client(2096): Receiver:  276.9 msg/s received      106.2 msg/s processed (38%)
Apr 07 11:18:45 ubuntu-noble piaware[1081]: mlat-client(2096): Server:      0.0 kB/s from server    0.0kB/s TCP to server     1.0kB/s UDP to server
Apr 07 11:18:45 ubuntu-noble piaware[1081]: mlat-client(2096): Aircraft: 12 of 23 Mode S, 33 of 45 ADS-B used
Apr 07 11:19:14 ubuntu-noble piaware[1081]: 5655 msgs recv'd from dump1090-fa (1648 in last 5m); 5655 msgs sent to FlightAware
pi@ubuntu-noble:~$
```

**Comment by jprochazka** [comment]
That's good news. This being said Questing SHOULD be good to go as well, being it is more inline with Trixie using Python 3.13.x.

**FYI as of today, the dev branch contained within the official repository is being used when Noble, Questing or Trixie is selected.**

**Comment by jprochazka** [comment]
Should be fixed with the same fix applied for #639.
